src/cpp/core/Database.cpp

```diff
@@ -0,0 +1,638 @@
+                          options.host %
+                          options.port %
+                          options.database %
+                          options.user %
+                          options.password %
+                          safe_convert::numberToString(options.connectionTimeoutSeconds, "0"));
+
+         boost::shared_ptr<IConnection> pConnection(new Connection(soci::postgresql, connectionStr));
+         *pPtrConnection_ = pConnection;
+         return Success();
+      }
+      catch (soci::soci_error& error)
+      {
+         return DatabaseError(error);
+      }
+   }
+
+private:
+   boost::shared_ptr<IConnection>* pPtrConnection_;
+};
+
+Query::Query(const std::string& sqlStatement,
+             soci::session& session) :
+   statement_(session)
+{
+   // it's possible that prepare can throw a database exception, but we
+   // do not want to surface errors until execute() is called
+   try
+   {
+      statement_.alloc();
+      statement_.prepare(sqlStatement);
+   }
+   catch (soci::soci_error& error)
+   {
+      prepareError_ = error;
+   }
+}
+
+RowsetIterator Rowset::begin()
+{
+   if (query_)
+      return RowsetIterator(query_.get().statement_, row_);
+
+   return end();
+}
+
+RowsetIterator Rowset::end()
+{
+   return RowsetIterator();
+}
+
+Connection::Connection(const soci::backend_factory& factory,
+                       const std::string& connectionStr) :
+   session_(factory, connectionStr)
+{
+}
+
+Query Connection::query(const std::string& sqlStatement)
+{
+   return Query(sqlStatement, session_);
+}
+
+Error Connection::execute(Query& query,
+                          bool* pDataReturned)
+{
+   if (query.prepareError_)
+      return DatabaseError(query.prepareError_.get());
+
+   try
+   {
+      query.statement_.define_and_bind();
+      bool result = query.statement_.execute(true);
+
+      if (pDataReturned)
+         *pDataReturned = result;
+
+      query.statement_.bind_clean_up();
+
+      return Success();
+   }
+   catch (soci::soci_error& error)
+   {
+      return DatabaseError(error);
+   }
+}
+
+Error Connection::execute(Query& query,
+                          Rowset& rowset)
+{
+   if (query.prepareError_)
+      return DatabaseError(query.prepareError_.get());
+
+   try
+   {
+      query.statement_.define_and_bind();
+      query.statement_.exchange_for_rowset(soci::into(rowset.row_));
+      query.statement_.execute(false);
+
+      rowset.query_ = query;
+
+      return Success();
+   }
+   catch (soci::soci_error& error)
+   {
+      return DatabaseError(error);
+   }
+}
+
+Error Connection::executeStr(const std::string& queryStr)
+{
+   try
+   {
+      // SOCI backends do not necessarily support running multiple statements
+      // in one invocation - to work around this, we split any passed in SQL
+      // into one invocation per SQL statement (delimited by ;)
+      std::vector<std::string> queries;
+      boost::regex regex(";[ \\t\\r\\f\\v]*\\n");
```

Comment:
Correct, this explicitly will not split on that, it will split on only the very last `;`, because the regex requires newline after the `;`.

Reply:
Ah, I think that makes sense since we'll never get a full query (or set of multiple queries) from the user directly - just values for filtering or inserting, right? Can it (or does it need to) support `;`  followed by `\\n` within the literal filter data?

Reply:
I really don't think we need to support that. This is only for schema update queries - it's not ever going to be 100% perfect (without vastly overcomplicating this code). I think as written is good enough - this isn't user hardened, it only needs to work for our particular crafted queries in our migration scripts, which I think should pretty much have us covered 100%.